File: InterfaceBuilding/SwiftInterfaceBulding/SwiftInterfaceBulding/SwapTester.swift
//
//  SwapTester.swift
//  SwiftInterfaceBulding
//
//  Created by Krzysztof Drab on 29/12/2020.
//  Copyright © 2020 Krzysztof Drab. All rights reserved.
//

import Foundation

class SwapTester: Tester {
    private var nbOfItems: Int = 0
    private var nbOfChanges: Int = 0
    private var rows: [RowData] = []
    
    override func generateRows() -> [RowData] {
        self.tests = [5, 25, 50, 75, 100, 125, 150, 175, 200];
        self.nbOfItems = 800;
        self.nbOfChanges = tests[testId];
        self.rows = self.generateDefaultRows(nbOfRows: self.nbOfItems);
        return self.rows;
    }

    
    override func updateRows() -> [RowData] {
        let startingPoint = Date()
        var itemsToChange = self.nbOfChanges;
        while(itemsToChange > 0) {
            let i = Int.random(in: 0..<self.rows.count);
            let j = Int.random(in: 0..<self.rows.count);
            rows.swapAt(i, j);
            itemsToChange -= 1;
        }
        self.timeToModify += startingPoint.timeIntervalSinceNow * -1;
        return self.rows;
    }
    
    override func testEnded(fps: Double) {
        self.results.append(TestResult(n: nbOfChanges, timeToModify: timeToModify, fps: fps))
    }
}

File: InterfaceBuilding/flutter_interface_building/lib/swap_tester.dart
import 'dart:io' show Platform;
import 'dart:math';

import 'row_data.dart';
import 'tester.dart';

class SwapTester extends Tester {
  List<int> tests = Platform.isIOS
    ? [5, 25, 50, 75, 100, 125, 150, 175, 200]
    : [1, 10, 20, 30, 40, 50, 60, 70, 80];
  int nbOfItems = Platform.isIOS ? 800 : 150;
  int nbOfChanges = 0;
  List<RowData> rows;

  List<RowData> generateRows() {
    this.nbOfChanges = this.tests[this.testId];
    this.rows = this.generateDefaultRows(this.nbOfItems);
    return this.rows;
  }

  List<RowData> updateRows() {
    final stopwatch = Stopwatch()..start();
    var rng = new Random();
    RowData temp;
    for(var i = 0 ; i < this.nbOfChanges ; i++) {
      int a = rng.nextInt(this.nbOfItems);
      int b = rng.nextInt(this.nbOfItems);
      temp = this.rows[a];
      this.rows[a] = this.rows[b];
      this.rows[b] = temp;
    }
    this.timeToModify += stopwatch.elapsedMilliseconds;
    return this.rows;
  }

  void testEnded(double fps) {
    this.results.add(TestResult(this.nbOfChanges, fps, this.timeToModify));
  }
}